### PR TITLE
[TunableOp] Size screening profiles by the tuning limits

### LINKED ISSUE #190433: Heavy op not handled gracefully by TunableOp
### 🐛 Describe the bug

This bug is relayed from [ROCm/TheRock](https://github.com/ROCm/TheRock/issues/5550).
Instead of stop tuning after the set duration or iterations, TunableOp just continues.
Check the original issue for more details.
To reproduce:
```
import torch
import os

os.environ["PYTORCH_TUNABLEOP_ENABLED"] = "1"
os.environ["PYTORCH_TUNABLEOP_TUNING"] = "1"
os.environ["PYTORCH_TUNABLEOP_VERBOSE"] = "3"
os.environ["PYTORCH_TUNABLEOP_HIPBLASLT_ENABLED"] = "0" # Have to be false for gfx906 or rocblaslt error
os.environ["PYTORCH_TUNABLEOP_FILENAME"] = "tuned_gemm.csv"
os.environ["PYTORCH_TUNABLEOP_MAX_TUNING_DURATION_MS"] = "50"
os.environ["PYTORCH_TUNABLEOP_MAX_TUNING_ITERATIONS"] = "5"

print(torch.cuda.tunable.get_max_tuning_duration())
print(torch.cuda.tunable.get_max_tuning_iterations())

M, N, K = 65536, 32768, 384
dtype = torch.float16
device = torch.device("cuda")

A = torch.randn(M, K, dtype=dtype, device=device)
B = torch.randn(N, K, dtype=dtype, device=device)
C = torch.empty(M, N, dtype=dtype, device=device)

torch.mm(A, B.T, out=C)

```

### Versions

```
$ python collect_env.py
Collecting environment information...
PyTorch version: 2.12.0+rocm7.15.0a20260718
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 7.15.0

OS: Linux Mint 21.3 (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.13.14 (main, Jun 11 2026, 12:30:58) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-110-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: AMD Radeon VII (gfx906:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: 7.15.0
MIOpen runtime version: 3.5.2
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
架构：                                   x86_64
CPU 运行模式：                           32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
字节序：                                 Little Endian
CPU:                                     12
在线 CPU 列表：                          0-11
厂商 ID：                                AuthenticAMD
型号名称：                               AMD Ryzen 5 5500
CPU 系列：                               25
型号：                                   80
每个核的线程数：                         2
每个座的核数：                           6
座：                                     1
步进：                                   0
Frequency boost:                         enabled
CPU 最大 MHz：                           4359.0000
CPU 最小 MHz：                           400.0000
BogoMIPS：                               7186.38
标记：                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm debug_swap ibpb_exit_to_user
L1d 缓存：                               192 KiB (6 instances)
L1i 缓存：                               192 KiB (6 instances)
L2 缓存：                                3 MiB (6 instances)
L3 缓存：                                16 MiB (1 instance)
NUMA 节点：                              1
NUMA 节点0 CPU：                         0-11
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] amd-torch-device-gfx906==2.12.0+rocm7.15.0a20260718
[pip3] amd-torchvision-device-gfx906==0.27.0+rocm7.15.0a20260718
[pip3] numpy==2.4.4
[pip3] torch==2.12.0+rocm7.15.0a20260718
[pip3] torchaudio==2.11.0+rocm7.15.0a20260718
[pip3] torchvision==0.27.0+rocm7.15.0a20260718
[pip3] triton==3.7.1+git0263a6a6.rocm7.15.0a20260718
[conda] Could not collect
```

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### PR DESCRIPTION
**Pull request by naromero77amd**
Fixes #190433

## Problem

TunableOp screens every candidate twice before profiling it for real, at a hard-coded 3 and then 10 iterations. Those screens exist only to avoid the cost of the final tuning profile, but their cost is a fixed constant while the profile they protect is derived from the configured limits. The two diverge exactly when calls are expensive: on the GEMM from #190433, with `PYTORCH_TUNABLEOP_MAX_TUNING_DURATION_MS=50` against a 34 ms call, the final profile is capped at a single iteration while the screens still spend 17 calls to save at most 3.

The defect is therefore not that the limits are ignored, but that the screening passes are blind to them.

## Change

`ScreenIters()` applies the same limits that already size the final tuning profile, treating the 3 and 10 as caps rather than constants, so a screen can never cost more than the profile it protects. The second screen runs only when the limits leave room for more iterations than the first one got, since below that it cannot improve on the estimate already in hand. Sizing the first screen needs an estimate before any profiling has happened, so the support probe that already runs for every candidate is now timed; it carries one-time setup cost and therefore over-estimates, which is the safe direction when the value is used only to bound work.

`ProfileStats` also ran two mandatory warmup calls on every invocation, including on a candidate it had just finished profiling. Warmup now happens only on a candidate's first touch, which removes four redundant calls per candidate independently of any limit.

Separately, `StreamTimer` and `StreamTimerNoSync` each created two CUDA events and never destroyed them, leaking `2 * num_iter` events per `ProfileStats` call and roughly 200 per candidate at default settings. Both destructors now release their events, and copy and move are deleted because a user-provided destructor would otherwise leave the deprecated implicit copy constructor in place on a type held by `std::vector`.

## Relationship to the first commit

The first commit on this branch skipped the second screen outright whenever a limit was small. That is simpler but discontinuous, and at the shipped defaults of 30 ms and 100 iterations it disables the 1.15x filter for every op slower than roughly 3 ms, which is a large class of real GEMMs. The second commit replaces it with the scaled version, which keeps the filter wherever the budget can still fund it and reaches zero only when it genuinely cannot.

## Results

Tuning wall time for a single `torch.mm`, measured on gfx90a with 4320 candidates per shape:

| shape (per-call) | main | 1st commit | this PR |
|---|---|---|---|
| 1024^3 (0.036 ms) | 3.14 s | 3.17 s | 3.13 s |
| 4096^3 (1.087 ms) | 37.97 s | 37.94 s | 35.41 s |
| 8192x8192x4096 (4.145 ms) | 124.56 s | 104.98 s | 94.67 s |
| 65536x32768x384, 50/5 limits (34 ms) | 560.14 s | 431.40 s | 274.14 s |

Verbose logs confirm the mechanism, not just the outcome. For the 4.145 ms shape at default limits the second screen is retained and stays equally selective (307 candidates killed on main, 313 here), while the first commit drops it entirely (0 killed, and the population reaching full tuning goes from 289 to 596). For the 34 ms shape the screen correctly collapses to nothing in both.

Tuned solution quality is unchanged. Recorded winner times move from 0.9460 to 0.9458 ms, 3.6125 to 3.6168 ms, and 21.5442 to 21.5453 ms. Two shapes select a different solution ID among near-equivalent candidates with no measurable throughput difference.

## Test plan

```
python test/test_linalg.py -v -k tunableop
```

41 tests, OK with 12 platform skips (gfx942-only TF32, FP8 MI300+, ROCm-unsupported cuBLASLt cases) in 150.8s on gfx90a.

A positive control was run before any measurement, forcing `max_tuning_iterations=2` on a cheap GEMM and confirming zero second-screen lines and `tuning iters 2` in the verbose log, so the numbers above are not from a stale hipified build.

Validation is ROCm-only; there is no NVIDIA hardware on the machine used, so the cuBLASLt path is left to upstream CI.

## Scope

What remains after this change is the candidate count multiplied by the per-candidate cost, and a single fp16 GEMM enumerates 4320 candidates on this hardware. Bounding that total would mean abandoning the sweep partway through, which the existing knobs deliberately do not do: `PYTORCH_TUNABLEOP_MAX_TUNING_DURATION_MS` and `PYTORCH_TUNABLEOP_MAX_TUNING_ITERATIONS` are per-solution by documented design. A total budget would be a new user-facing knob whose result depends on enumeration order and machine load rather than being reproducible, so it is a separate proposal on its own merits, not part of this fix.

What was genuinely defective is the disproportion, screening that costs several times what the profile it protects is permitted to cost. That is what this fixes, and it is what we can reasonably achieve without changing the meaning of the existing knobs: 2x on the reported case, no loss of tuned solution quality, and no behavior change for ops cheap enough that the limits never bind.

## AI assistance

> The second commit and this description were drafted with Claude (Claude Code). The design, the measurements, and the conclusions were reviewed by the author before posting.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Comment by naromero77amd** [comment]
@claude look at the CI logs for ROCm did all the tunablop unit test in the test_linalg unit test suite run. Did all the unit test run and did any of them take more than 2 minutes to run to completion?

**Review by jeffdaily — APPROVED** [review]
